[WIP] prépare la migration vers zmd

vhf opened the pull request:
| Q                                   | R
| ----------------------------------- | -------------------------------------------
| Type de modification                | évolution
| Ticket(s) (_issue(s)_) concerné(s)  | #4388

Cette PR prépare le terrain pour la migration vers zmarkdown. Le but est de voir comment les choses fonctionnent actuellement, de documenter le travail qui sera nécessaire à la migration, et ce faisant de passer un coup de balais.

vhf committed:
WIP: prépare la migration vers zmd

vhf commented:
Pour les curieux : 

* Server: https://github.com/vhf/zds-site/blob/5092aba14f7b1d7fe368da27c25b97f3762ab2c2/zmarkdown/markdown-server.js
* Client: https://github.com/vhf/zds-site/blob/c690e049e77d4b65dfe8c6612dfc58e061324823/zds/utils/templatetags/emarkdown.py#L35-L36

Test manuel / demo :

```py
import zerorpc

markdown_client = zerorpc.Client()
markdown_client.connect("tcp://127.0.0.1:24242")

html = markdown_client.toHTML("""
## My title

Foo: [bar](http://example.com)
""")

print html
```

```
$ yarn
$ npm run zmd
$ python zmd-server-test.py
<h2>My title</h2>
<p>Foo: <a href="http://example.com">bar</a></p>
```

artragis commented on zmarkdown/config.js:
```diff
@@ -0,0 +1,322 @@
+  },
+}
+
+const rebberConfig = {
```
remember we need to override `image` and `figure.image` to use the custome `\\image` command in rebber.

vhf replied:
Oui ça viendra :)

Là c'est juste pour que ça fonctionne temporairement. Quand tout est bien câblé on verra où et comment on configure le serveur de rendu markdown/latex.

artragis committed:
prepare epub_work

vhf committed:
zmarkdown: remplace zerorpc par un serveur http

SpaceFox commented on zds/utils/templatetags/emarkdown.py:
```diff
@@ -22,35 +21,36 @@
     inline = kwargs.get('inline', False) is True
     is_latex = kwargs.pop('is_latex', False) is True
-    markdown_compiler = markdown_client.toHTML if not is_latex else markdown_client.toLatex
+    endpoint = '/html' if not is_latex else '/latex'
```
Le Comité Contre les Double Négations préfèrerait – si c'est possible – un truc du genre :

```python
endpoint = '/latex' if is_latex else 'html'
```

À voir selon les habitudes du langage aussi.

vhf replied:
Ouais peut-être, à voir.

Je me serais pas permis un `x = y if not not_foo else foo`, ça ce serait une double négation et ce serait pénible à lire. J'ai fait ça comme ça ici parce que clairement 99% des requêtes iront sur l'endpoint `/html` et que je préfère noter que sauf exception c'est là qu'on va.

Je crois que je préférerais

```py
endoint = '/html'
if is_latex:
    endpoint = '/latex'
```

à ta solution, pour continuer à véhiculer cette impression de valeur par défaut.

artragis replied:
Lorsque j'avais fait la condition initiale j'avais pensé justement au fait que plus de 90% des requêtes seraient vers html. Après mettez la forme que vous voulez, les deux sont correctes et lisibles.

Commit:
Debug generation.

vhf committed:
Upgrade zmarkdown

vhf committed:
Ajoute quelques notes pour la suite

vhf committed:
zmarkdown: ajoute le template document latex

vhf committed:
Style fixes

vhf commented on .travis.yml:
```diff
@@ -86,6 +87,9 @@ install:
   - export PATH=$PATH:$HOME/bin
+  - mkdir -p $HOME/.texlive/texmf-local/tex/latex/zestedesavoir
+  - cp -f $HOME/.cache/yarn/npm-zmarkdown*/zmdocument.cls $HOME/.texlive/texmf-local/tex/latex/zestedesavoir
+  - $HOME/.texlive/bin/x86_64-linux/mktexlsr $HOME/.texlive
```
@pierre-24 @Karnaj une idée de comment faire pour que LaTeX trouve la classe ?

Je la mets là : `$HOME/.texlive/texmf-local/tex/latex/zestedesavoir/zmdocument.cls` mais si vous regardez les logs de travis, latex ne trouve pas la classe.

pierre-24 replied:
J'imagine que t'as déjà essayé, mais chez moi, je met simplement le .cls dans `$HOME/texmf/tex/latex/`, puis je fais un `texhash` et ça fonctionne sans autre forme de procès.

pierre-24 replied:
... Voilà, [ici](https://travis-ci.org/pierre-24/latex-template/builds/259902363?utm_source=email&utm_medium=notification), j'ai réussi. Ton "erreur", c'est le "zestedesavoir", du reste t'étais quasi dans le bon :)

vhf replied:
Fantastique, merci pour ton aide !

Maintenant il doit nous manquer des trucs :

>     FailureDuringPublication: ! LaTeX Error: File xpatch.sty not found.

Je remets un message ici si je trouve

vhf replied:
Réussi !

artragis committed:
fix template

vhf committed:
Tests: supprime 2 références à Pandoc

Nécessaire pour que les tests passent. Il reste beaucoup de références à enlever
mais rien d'indispensable pour les tests.

vhf committed:
Chemin vers les smileys

artragis committed:
fix zip

artragis committed:
fix markdown-server
so that rebber returns a good latex string
sorry I could not prevent this joke.

artragis committed:
use xelatex for utf8 purpose

artragis committed:
First attempt for ebook

sandhose committed:
Feuille de style séparée pour le ZMarkdown

sandhose committed:
Correction du chemin vers le CSS des EPUBs

Pas testé.

motet-a commented on zds/tutorialv2/epub_utils.py:
```diff
@@ -0,0 +1,98 @@
+
+
+def build_toc_cnx(chapters, tutorial, working_dir):
+    with open(join(working_dir, 'toc.ncx'), mode='wb', encoding='utf-8') as f:
```
cnx ou ncx ?

vhf committed:
Update lockfile

vhf committed:
Adapte travis suite au rebase

vhf committed:
Règle un import django.conf

motet-a commented on Gulpfile.js:
```diff
@@ -91,8 +91,8 @@ gulp.task('js', () =>
 // Compiles the SCSS files to CSS
-gulp.task('css', ['css:sprite'], () =>
-    gulp.src('assets/scss/main.scss')
+gulp.task('css', ['css:sprite', 'css:vendors'], () =>
```
La tâche `css:vendors` n’existe plus désormais. Les dépendances externes (`normalize.css`) sont maintenant accédées directement depuis le `node_modules`.

motet-a commented on Gulpfile.js:
```diff
@@ -91,8 +91,8 @@ gulp.task('js', () =>
-gulp.task('css', ['css:sprite'], () =>
-    gulp.src('assets/scss/main.scss')
+gulp.task('css', ['css:sprite', 'css:vendors'], () =>
+    gulp.src(['assets/scss/main.scss', 'assets/scss/zmd.scss'])
```
Pourquoi ne pas `@import` le fichier `zmd.scss` depuis `main.scss` ?

vhf committed:
Enlève une tâche gulp qui n'existe plus

vhf committed:
Installe les dépendances pour zmd

vhf committed:
Fix les feuilles de style

vhf committed:
Supprime classe CSS inutilisée wf-active

vhf committed:
zmarkdown: extrait les controlleurs express

vhf committed:
zmarkdown: upgrade

vhf committed:
Fix a few tests

artragis commented on zmarkdown/markdown-server.js:
```diff
@@ -1,14 +1,20 @@
 const app = express()
+
+app.use(Raven.requestHandler())
+app.use(Raven.errorHandler());
```
ça peut être cool de trouver la version de zmd pour ajouter ça aux release management.

vhf committed:
Améliore le reporting sentry node

artragis commented on zmarkdown/markdown-server.js:
```diff
@@ -1,30 +1,41 @@
+  Raven.config(process.env.SENTRY_DSN).install();
 
-const app = express()
+  Raven.mergeContext({
```
il existe la propriété "release" qui peut être sympa. mais là ça peut suffire, puisque les tags sont indexés.

vhf replied:
Je veux bien un lien vers la doc, j'ai cherché et j'ai aussi comparé avec ce qu'on fait avec Django. :)

vhf replied:
Ah je crois que c'était bien caché : https://docs.sentry.io/clients/javascript/sourcemaps/

Si je comprends bien je dois modifier le DSN pour y ajouter `/releases/<sha>` et passer

```
Raven.config({
    release: '<sha>'
})
```

vhf committed:
Memoize séparément les compilos html et latex

artragis commented on zds/tutorialv2/publication_utils.py:
```diff
@@ -238,47 +232,180 @@ def publish(self, md_file_path, base_name, **kwargs):
+        md_flat_content = _read_flat_markdown(md_file_path)
+        published_content_entity = self.get_published_content_entity(md_file_path)
+        depth_to_size_map = {
+            2: 'small',
```
tu peux ajouter 1: 'small' ici.

vhf commented on requirements.txt:
```diff
@@ -36,3 +36,4 @@ watchdog==0.8.3 # use last non-bc-breaking version
 # next versions of this libraries break previous behavior for slug generation from string with single quotes
 django-uuslug==1.0.3
 python-slugify==1.1.4
+Pygments==2.2.0
```
Note: faudra faire tourner latex dans le venv pour qu'il puisse accéder à Pygments.

vhf committed:
Upgrade zmarkdown

vhf commented:
Suite ici comme ça tout le monde peut contribuer : #4617
